sky/utils/vpn_utils.py

```diff
@@ -0,0 +1,285 @@
+
+import requests
+
+from sky import sky_logging
+from sky.provision import common
+from sky.utils import common_utils
+from sky.utils import schemas
+from sky.utils import subprocess_utils
+
+logger = sky_logging.init_logger(__name__)
+
+
+class VPN:
+    """Interface for VPN configuration."""
+
+    def get_setup_command(self) -> str:
+        """Get the command to setup the VPN on VM instances."""
+        raise NotImplementedError
+
+    def get_setup_env_vars(self) -> Dict[str, str]:
+        """Get the environment variables to setup instances with the VPN.
+
+        NOTE: This is used for launching another instance with the same VPN
+        by SkyPilot, instead of setting up the VPN on that instance. Currently
+        it is only used in SkyServe.
+        """
+        return {}
+
+    def get_private_ip(self, hostname: str) -> str:
+        """Get the private IP address from the hostname."""
+        raise NotImplementedError
+
+    def remove_host(self, hostname: str) -> None:
+        """Remove a host from the VPN."""
+        raise NotImplementedError
+
+    def to_yaml_config(self) -> Dict[str, Any]:
+        """Get the VPN configuration in YAML format."""
+        return {}
+
+    @staticmethod
+    def from_backend_config(config: Dict[str, Any]) -> 'VPN':
+        """Create a VPN configuration from the backend configuration."""
+        raise NotImplementedError
+
+    def to_backend_config(self) -> Dict[str, Any]:
+        """Get the VPN configuration for the backend."""
+        return {}
+
+
+class TailscaleVPN(VPN):
+    """Tailscale VPN configuration."""
+
+    # pylint: disable=line-too-long
+    _SETUP_COMMAND = (
+        'curl -fsSL https://pkgs.tailscale.com/stable/ubuntu/focal.noarmor.gpg | '
+        'sudo tee /usr/share/keyrings/tailscale-archive-keyring.gpg >/dev/null; '
+        'curl -fsSL https://pkgs.tailscale.com/stable/ubuntu/focal.tailscale-keyring.list | '
+        'sudo tee /etc/apt/sources.list.d/tailscale.list >/dev/null; '
+        'sudo apt-get update > /dev/null 2>&1; '
+        'sudo apt-get install tailscale -y > /dev/null 2>&1; '
+        'sudo tailscale login --auth-key {tailscale_auth_key}')
+
+    def __init__(
+        self,
+        auth_key: str,
+        api_key: Optional[str] = None,
+        network_name: Optional[str] = None,
+        enable_api: bool = False,
+    ) -> None:
+        self._auth_key = auth_key
+        self._api_key = api_key
+        self._network_name = network_name
+        self._enable_api = enable_api
+        if not enable_api:
+            logger.warning('TAILSCALE_API_KEY or TAILSCALE_NETWORK_NAME is'
+                           ' not set. Tailscale API will be disabled. You'
+                           ' may need to remove hosts manually.')
+        elif not api_key or not network_name:
+            raise ValueError('Both TAILSCALE_API_KEY and TAILSCALE_NETWORK_NAME'
+                             ' must be set to enable the Tailscale API.')
```

Comment:
```suggestion
            raise ValueError('Both TAILSCALE_API_KEY and TAILSCALE_NETWORK_NAME '
                             'must be set to enable the Tailscale API.')
```
nit: put the space on the end of precious line. pls check other places in your code

Reply:
Also pls use this:https://github.com/skypilot-org/skypilot/blob/b1f22c4d5fe0a3cc25d1df0a8a05d4230a28b702/sky/utils/ux_utils.py#L26-L40